nexus/db-queries/src/db/datastore/deployment.rs

```diff
@@ -0,0 +1,1578 @@
+                };
+
+                Ok((
+                    nblueprints,
+                    nsled_agent_zones,
+                    nzones,
+                    nnics,
+                    nzones_not_in_service,
+                ))
+            })
+            .await
+            .map_err(|error| match error {
+                TransactionError::CustomError(e) => e,
+                TransactionError::Database(e) => {
+                    public_error_from_diesel(e, ErrorHandler::Server)
+                }
+            })?;
+
+        info!(&opctx.log, "removed blueprint";
+            "blueprint_id" => blueprint_id.to_string(),
+            "nblueprints" => nblueprints,
+            "nsled_agent_zones" => nsled_agent_zones,
+            "nzones" => nzones,
+            "nnics" => nnics,
+            "nzones_not_in_service" => nzones_not_in_service,
+        );
+
+        Ok(())
+    }
+
+    /// Set the current target blueprint
+    ///
+    /// In order to become the target blueprint, `target`'s parent blueprint
+    /// must be the current target
+    pub async fn blueprint_target_set_current(
+        &self,
+        opctx: &OpContext,
+        target: BlueprintTarget,
+    ) -> Result<(), Error> {
+        opctx
+            .authorize(authz::Action::Modify, &authz::BLUEPRINT_CONFIG)
+            .await?;
+
+        let query = InsertTargetQuery {
+            target_id: target.target_id,
+            enabled: target.enabled,
+            time_made_target: target.time_made_target,
+        };
+
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        query
+            .execute_async(&*conn)
+            .await
+            .map_err(|e| Error::from(query.decode_error(e)))?;
+
+        Ok(())
+    }
+
+    /// Get the current target blueprint, if one exists
+    ///
+    /// Returns both the metadata about the target and the full blueprint
+    /// contents. If you only need the target metadata, use
+    /// `blueprint_target_get_current` instead.
+    pub async fn blueprint_target_get_current_full(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Option<(BlueprintTarget, Blueprint)>, Error> {
+        opctx.authorize(authz::Action::Read, &authz::BLUEPRINT_CONFIG).await?;
+
+        let conn = self.pool_connection_authorized(opctx).await?;
+        let Some(target) = self.blueprint_current_target_only(&conn).await?
+        else {
+            return Ok(None);
+        };
+
+        // The blueprint for the current target cannot be deleted while it is
+        // the current target, but it's possible someone else (a) made a new
+        // blueprint the target and (b) deleted the blueprint pointed to by our
+        // `target` between the above query and the below query. In such a case,
+        // this query will fail with an "unknown blueprint ID" error. This
+        // should be rare in practice.
+        //
+        // TODO-correctness Should we loop here if we get a "blueprint not
```

Comment:
Agreed.  I generally don't like retrying for stuff like this because it creates latency that's very non-transparent when we hit these cases (i.e., as we see contention, things just start getting slow, and it's hard for both clients and us to see why).  A 500 or 503 might better reflect that this is sort of our bad, but it's an ugly wart (but also doesn't seem worth fixing)...

Reply:
:+1: Removed the TODO